p4-16/spec/P4-16-spec.mdk

```diff
@@ -6416,6 +6416,160 @@ ck16.update( { h.ipv4.ttl, h.ipv4.proto } );
+parsers and controls.  Variables of these types may not be defined at
+the top level of your program.
+
+|-------------|-------------------------------------------------|
+|             | can be instantiated in this place |||||
+|             |-----------|---------|---------|---------|--------|
+| This type   | top level | package | parser  | control | extern |
++-------------+:---------:+:-------:+:-------:+:-------:+:------:+
+| package     | yes       |  no     | no      | no      |  no    |
+| parser      |  no       |  no     | yes     | no      |  no    |
+| control     |  no       |  no     | no      | yes     |  no    |
+| extern      | yes       |  no     | yes     | yes     |  no    |
+| table       |  no       |  no     | no      | yes     |  no    |
+| value types |  no       |  no     | yes     | yes     |  no    |
+|-------------|-----------|---------|---------|---------|--------|
+
+The next table lists restrictions on what types can be passed as
+run-time parameters to other callable things that have run-time
+parameters: parsers, controls, extern methods, and actions.
+
+|-------------|-------------------------------------------------|
+|             | can be a run-time parameter to this callable thing ||||
+|             |-----------|---------|---------|---------|
+| This type   | parser    | control | method  | action  |
++-------------+:---------:+:-------:+:-------:+:-------:+
+| package     |  no       |  no     | no      | no      |
+| parser      |  no       |  no     | no      | no      |
+| control     |  no       |  no     | no      | no      |
+| extern      | yes       | yes     | yes     | no      |
+| table       |  no       |  no     | no      | no      |
+| value types | yes       | yes     | yes     | yes     |
+|-------------|-----------|---------|---------|---------|
+
+The restrictions on return types of extern methods is the same as for
+parameters, with the addition that `void` can be a return type for
+methods that return no value.
+
+The next table lists restrictions on what kinds of calls can be made
+from which places in a P4 program.  Calling a parser, control, or
+table means invoking its `apply()` method.  The row for `extern`
+describes where extern method calls can be made from.
+
+|-------------|-------------------------------------------------|
+|             | can be called at run time from this place in a P4 program |||||
+|             |-----------|---------|---------|------------|--------|
+|             |           | control | action  | action not |        |
+|             | parser    | apply   | used by | used by    |        |
+| This type   | state     | block   | table   | any table  | extern |
++-------------+:---------:+:-------:+:-------:+:----------:+:------:+
+| package     | N/A       | N/A     | N/A     | N/A        | N/A    |
+| parser      | yes       |  no     | no      | no         | no     |
+| control     |  no       | yes     | no      | yes        | no     |
```

Comment:
I think that actions should not be allowed to call controls. So the fourth entry should be "no".
I understand that you are treating here actions as functions; perhaps we should rather change the language to add functions rather than carry this baggage with two kinds of actions going forward.

Note that top-level actions cannot call controls because they have no way of naming control instances; you can only call a control instance.